power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_se_solver.hpp

```diff
@@ -21,17 +22,443 @@ namespace power_grid_model {
+template <bool sym> struct NRSEUnknown : public Block<double, sym, false, 4> {
+    template <int r, int c> using GetterType = typename Block<double, sym, false, 4>::template GetterType<r, c>;
+
+    // eigen expression
+    using Block<double, sym, false, 4>::Block;
+    using Block<double, sym, false, 4>::operator=;
+
+    GetterType<0, 0> theta() { return this->template get_val<0, 0>(); }
+    GetterType<1, 0> v() { return this->template get_val<1, 0>(); }
+    GetterType<2, 0> phi_theta() { return this->template get_val<2, 0>(); }
+    GetterType<3, 0> phi_v() { return this->template get_val<3, 0>(); }
+
+    GetterType<0, 0> eta_theta() { return this->template get_val<0, 0>(); }
+    GetterType<1, 0> eta_v() { return this->template get_val<1, 0>(); }
+    GetterType<2, 0> tau_theta() { return this->template get_val<2, 0>(); }
+    GetterType<3, 0> tau_v() { return this->template get_val<3, 0>(); }
+};
+
+// block class for the right hand side in state estimation equation
+template <bool sym> using NRSERhs = NRSEUnknown<sym>;
+
+// class of 4*4 (12*12) se gain block
+/*
+[
+   [G, QH]
+   [Q, R ]
+]
+*/
+template <bool sym> class NRSEGainBlock : public Block<double, sym, true, 4> {
+  public:
+    template <int r, int c> using GetterType = typename Block<double, sym, true, 4>::template GetterType<r, c>;
+
+    // eigen expression
+    using Block<double, sym, true, 4>::Block;
+    using Block<double, sym, true, 4>::operator=;
+
+    GetterType<0, 0> g_P_theta() { return this->template get_val<0, 0>(); }
+    GetterType<0, 1> g_P_v() { return this->template get_val<0, 1>(); }
+    GetterType<1, 0> g_Q_theta() { return this->template get_val<1, 0>(); }
+    GetterType<1, 1> g_Q_v() { return this->template get_val<1, 1>(); }
+
+    GetterType<0, 2> qh_P_theta() { return this->template get_val<0, 2>(); }
+    GetterType<0, 3> qh_P_v() { return this->template get_val<0, 3>(); }
+    GetterType<1, 2> qh_Q_theta() { return this->template get_val<1, 2>(); }
+    GetterType<1, 3> qh_Q_v() { return this->template get_val<1, 3>(); }
+
+    GetterType<2, 0> q_P_theta() { return this->template get_val<2, 0>(); }
+    GetterType<2, 1> q_P_v() { return this->template get_val<2, 1>(); }
+    GetterType<3, 0> q_Q_theta() { return this->template get_val<3, 0>(); }
+    GetterType<3, 1> q_Q_v() { return this->template get_val<3, 1>(); }
+
+    GetterType<2, 2> r_P_theta() { return this->template get_val<2, 2>(); }
+    GetterType<2, 3> r_P_v() { return this->template get_val<2, 3>(); }
+    GetterType<3, 2> r_Q_theta() { return this->template get_val<3, 2>(); }
+    GetterType<3, 3> r_Q_v() { return this->template get_val<3, 3>(); }
+};
+
 // solver
 template <bool sym> class NewtonRaphsonSESolver {
+
+    struct NRSEJacobian {
+        struct i_side_block {};
+        struct j_side_block {};
+
+        RealTensor<sym> dP_dt{};
+        RealTensor<sym> dP_dv{};
+        RealTensor<sym> dQ_dt{};
+        RealTensor<sym> dQ_dv{};
```

Comment:
Maybe it whould be nice to be explicit here, i.e. use `theta` instead of `t`. What do you think?

```suggestion
        RealTensor<sym> dP_dtheta{};
        RealTensor<sym> dP_dv{};
        RealTensor<sym> dQ_dtheta{};
        RealTensor<sym> dQ_dv{};
```

Reply:
It looked good when all things were multiplying together as it aligned well :D

Reply:
blue pill/red pill 💀

Reply:
Okay, let's keep it `dt` then